CDAP-14619 use objects instead of JsonObject and JsonArray

albertshau opened the pull request:
Use proper java objects for the service endpoints instead of
manually manipulating json for many endpoints.
This makes the code much easier to understand and maintain.

albertshau committed:
CDAP-14619 use objects instead of JsonObject and JsonArray

Use proper java objects for the service endpoints instead of
manually manipulating json for many endpoints.
This makes the code much easier to understand and maintain.

CuriousVini commented on wrangler-service/pom.xml:
```diff
@@ -83,6 +78,11 @@
       <artifactId>commons-io</artifactId>
       <version>${commons.io.version}</version>
     </dependency>
+    <dependency>
```
Why is this change needed?

albertshau replied:
I started trying to remove this because cdap-etl-core is a pipeline internal module and should not be a dependency, but decided to do it separately and added it back.

I can move this back to where it was before

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java:
```diff
@@ -100,6 +101,7 @@ public String getContent(String charset) {
    * @param type class to converted to.
    * @return instance of type T as defined by the class.
    */
+  @Nullable
```
I am not able to comment there but [this method](https://github.com/data-integrations/wrangler/pull/284/files#diff-04ee9f099bc61a0c38a03a5107fbf374L67) also returns null values. We should add `Nullable` there too.

albertshau replied:
there are many warnings in this class, i'll fix them.

albertshau committed:
CDAP-14619 cleanup RequestExtractor class

The utility class contained several warnings. Annotated several
methods as nullable, removed an invalid cast, simplified boolean
returns, and removed unused methods.

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/ServiceResponse.java:
```diff
@@ -0,0 +1,55 @@
+  private final String message;
+  private final Integer count;
+  private final Collection<T> values;
+  // TODO: see if this is used by UI. It should be a boolean and not a string...
```
Do we need a jira to track this?

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/bigquery/BigQuerySpec.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Plugin specification for a BigQuery pipeline source.
+ *
+ * TODO: clean up this API. There is no reason for this class to exist.
```
same here. Any jira needed to track it?

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/ConnectionSample.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Information about a sample taken using a connection.
+ *
+ * TODO: check if these fields are all used/needed.
```
do we need jira for this?

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java:
```diff
@@ -56,14 +47,15 @@ public boolean hasHeader(String name) {
    * @return value defined for the header.
    */
-  public <T> T getHeader(String name, String defaultValue) {
+  public String getHeader(String name, String defaultValue) {
```
Can `defaultValue` be null?

albertshau replied:
if it's not annotated as nullable, callers shouldn't pass in null. In this particular method, there should be no reason to call this method passing in null as default, since it's the same as just getting the header.

That being said, there is a place in the code that passes in null. I'll change it to not call this method.

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/service/spanner/SpannerService.java:
```diff
@@ -241,15 +241,19 @@ public void specification(HttpServiceRequest request, final HttpServiceResponder
-        new PluginConfiguration<>("Spanner", "source", specification);
-
-      // creating map of name -> PluginConfiguration to be consistent with other services response format
-      Map<String, PluginConfiguration> pluginMap = ImmutableMap.of("Spanner", pluginConfiguration);
```
`PluginConfiguration` is no longer used anywhere?

albertshau replied:
nope, will remove it.

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java:
```diff
@@ -27,7 +27,7 @@
  * TestRecipe pojo that stores information about a recipe in the backend dataset.
```
`TestRecipe` - this should be updated

albertshau replied:
done

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java:
```diff
@@ -46,23 +46,16 @@
     this.description = description;
+    this.created = created;
+    this.updated = updated;
+    this.directives = GSON.toJson(directives);
```
previous implementation had directives as null if not specified. should we handle that case?

albertshau replied:
the services enforce that the directives always exist, so I changed this class to reflect that.

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeInfo.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Same as {@link RecipeDatum} except directives are stored as a list of strings.
+ *
+ * TODO: consolidate the classes when revamping backend storage
```
please add a jira number for it.

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeInfo.java:
```diff
@@ -0,0 +1,45 @@
+    this.created = created;
+    this.updated = updated;
+    this.directives = directives;
+    this.length = directives.size();
```
can directives be null?

albertshau replied:
no

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/service/recipe/RecipeService.java:
```diff
@@ -97,6 +96,8 @@
    *   "message" : "Successfully created recipe with id 'test'"
    * }
    *
+   * TODO: correct the REST semantics, this should be a POST
```
Please add a jira number for it.

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java:
```diff
@@ -183,56 +189,35 @@ public void exploreGCS(HttpServiceRequest request, HttpServiceResponder responde
-          }
-          object.addProperty("wrangle", isWrangeable);
-          values.add(object);
+          String fileType = detector.detectFileType(name.substring(name.lastIndexOf("/") + 1));
```
Can file type be unknown?

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java:
```diff
@@ -241,58 +226,37 @@ public void exploreGCS(HttpServiceRequest request, HttpServiceResponder responde
-            // We will not enable wrangling on unknown data.
-          }
-          object.addProperty("wrangle", isWrangeable);
+          String fileType = detector.detectFileType(blob.getName());
```
can fileType be unknown?

albertshau replied:
it cannot be null, but it can be the literal string 'unknown' if that's what you're asking

CuriousVini replied:
I see. Before `detectFileType` was throwing IOException which was setting fileType to unknown. I seems that method is no longer throwing IOException. So no need to handle it?

albertshau replied:
yeah, there are several methods that declare that they throw an exception that they can never throw, which I've cleaned up in a few places.

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/service/database/DatabaseService.java:
```diff
@@ -200,10 +201,12 @@ public String getPort() {
-      BufferedReader br = new BufferedReader(new InputStreamReader(is));
+    InputStream is = DatabaseService.class.getClassLoader().getResourceAsStream("drivers.mapping");
+    if (is == null) {
+      return;
```
should we log any error message?

albertshau replied:
done

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/ServiceUtils.java:
```diff
@@ -73,10 +71,8 @@ public static void error(HttpServiceResponder responder, String message) {
-    error.addProperty("status", HttpURLConnection.HTTP_NOT_FOUND);
-    error.addProperty("message", message);
-    sendJson(responder, HttpURLConnection.HTTP_NOT_FOUND, error.toString());
+    ServiceResponse<Void> response = new ServiceResponse<>(message);
```
can we just use error() method with status code notFound?

albertshau replied:
done

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSBucketInfo.java:
```diff
@@ -0,0 +1,48 @@
+    this.created = created;
+    this.wrangle = wrangle;
+    // this is constant, but is part of the response so must be a field in this class
+    this.directory = true;
```
If this value is always true, should it be there in the response?

albertshau replied:
This is part of the API and must be preserved because the same endpoint returns different objects depending on the request params.

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java:
```diff
@@ -0,0 +1,56 @@
+  private final String bucket;
+  private final String path;
+  private final String blob;
+  private final Long generation;
```
why this is `Long` can it be `long`?

albertshau replied:
it can be null, i'll annotate as such

CuriousVini commented on wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java:
```diff
@@ -0,0 +1,56 @@
+  private final String path;
+  private final String blob;
+  private final Long generation;
+  private final Long created;
```
should we mark all the nullable fields as `Nullable`?

albertshau replied:
We usually annotate methods, not the actual variable

CuriousVini reviewed (commented):
Couple of comments.
- should we also use objects instead of jsonobject in services like DirectivesService, KafkaService, Explorer, SchemaRegistryService?
- Please file jiras for TODOs mentioned in the java doc.
- Also please update the copyright year.

albertshau committed:
CDAP-14619 review cleanup

modified copyrights, added jiras to TODO comments, removed an
unused class and cleaned up some null related logic.

albertshau commented:
KafkaService and Explorer are not using JsonObject anymore. DirectiveService I'm saving for later because I couldn't easily figure out what their responses are, since they merge json objects and stuff like that.

I'll do SchemaRegistryService if we touch it, I'm not sure where it's being used.

CuriousVini reviewed (approved):
just one question. Code changes LGTM 👍
